libraries/primitive_consul_rack_lock.rb

```diff
@@ -0,0 +1,80 @@
+require_relative 'primitive_consul_lock'
+
+module Choregraphie
+  # This primitive is based on optimistic concurrency (using compare-and-swap) rather than consul sessions.
+  # It allows to support the unavailability of the local consul agent (for reboot, reinstall, ...)
+  # It is based on the notion of rack: all nodes inside a rack can take the lock at the same time. We only allow a given number of racks to take the lock at the same time.
+  # Note: this lock does not guarantee that all nodes of a given rack will be done at the same time.
+  class ConsulRackLock < ConsulLock
+    def initialize(options = {}, &block)
+      super(options) # don't pass the block since we'll call it later
+      # id of the rack
+      validate!(:rack, String)
+
+      if @options[:service]
+        raise ArgumentError, "`service` option is not supported for consul_rack_lock"
+      end
+
+      # this block could be used to configure diplomat if needed
+      yield if block_given?
+    end
+
+    def register(choregraphie)
+
+      choregraphie.before do
+        wait_until(:enter) { semaphore.enter(name: @options[:rack], server: @options[:id]) }
+      end
+
+      choregraphie.finish do
+        # hack: We can ignore failure there since it is only to release
+        # the lock. If there is a temporary failure, we can wait for the
+        # next run to release the lock without compromising safety.
+        # The reason we have to be a bit more relaxed here, is that all
+        # chef run including a choregraphie with this primitive try to
+        # release the lock at the end of a successful run
+        wait_until(:exit, max_failures: 5) { semaphore.exit(name: @options[:rack], server: @options[:id]) }
```

Comment:
This is a bit sad that you have to reimplement the whole register method, just because of the options mapping.
using `:rack` and `:id` in your implementation of `enter` and `exit` would have save you this.